src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -682,6 +688,126 @@ int32_t CryptoNative_X509StoreCtxCommitToChain(X509_STORE_CTX* storeCtx)
     return 1;
 }
 
+int32_t CryptoNative_X509StoreCtxResetForSignatureError(X509_STORE_CTX* storeCtx, X509_STORE** newStore)
+{
+    if (storeCtx == NULL || newStore == NULL)
+    {
+        return -1;
+    }
+
+    *newStore = NULL;
+
+    int errorDepth = X509_STORE_CTX_get_error_depth(storeCtx);
+    X509Stack* chain = X509_STORE_CTX_get0_chain(storeCtx);
+    int chainLength = sk_X509_num(chain);
+    X509_STORE* store = X509_STORE_CTX_get0_store(storeCtx);
+
+    // If the signature error was reported at the last element
+    if (chainLength - 1 == errorDepth)
+    {
+        X509* root;
+        X509* last = sk_X509_value(chain, errorDepth);
+
+        // If the last element is in the trust store we need to build a new trust store.
+        if (X509_STORE_CTX_get1_issuer(&root, storeCtx, last))
+        {
+            if (root == last)
+            {
+                // We know it's a non-zero refcount after this because last has one, too.
+                // So go ahead and undo the get1.
+                X509_free(root);
+
+                X509_STORE* tmpNew = X509_STORE_new();
+
+                if (tmpNew == NULL)
+                {
+                    return 0;
+                }
+
+                X509* duplicate = X509_dup(last);
+
+                if (duplicate == NULL)
+                {
+                    X509_STORE_free(tmpNew);
+                    return 0;
+                }
+
+                if (!X509_STORE_add_cert(tmpNew, duplicate))
+                {
+                    X509_free(duplicate);
+                    X509_STORE_free(tmpNew);
+                    return 0;
+                }
+
+                *newStore = tmpNew;
+                store = tmpNew;
+                chainLength--;
+            }
+            else
+            {
+                // This really shouldn't happen, since if we could have resolved it now
+                // it should have resolved during chain walk.
+                //
+                // But better safe than sorry.
+                X509_free(root);
```

Comment:
If we do the X509_free(root) as the first thing in both the if and the else, couldn't we just move it to before the if and eliminate the else entirely?

Reply:
We.... could.  But I was a bit hesitant to let a flow exist where we downref root to 0 (when it doesn't match) but we end up talking about it still.